inc/functions.php

```diff
@@ -566,47 +624,73 @@ function my_date($format, $stamp=0, $offset="", $ty=1, $adodb=false)
  */
 function my_mail($to, $subject, $message, $from="", $charset="", $headers="", $keep_alive=false, $format="text", $message_text="", $return_email="")
 {
-	global $mybb;
+	global $mybb, $plugins;
 	static $mail;
 
-	// Does our object not exist? Create it
-	if(!is_object($mail))
-	{
-		require_once MYBB_ROOT."inc/class_mailhandler.php";
+	// Get our mail handler.
+	$mail = &get_mybb_mailhandler();
 
-		if($mybb->settings['mail_handler'] == 'smtp')
-		{
-			require_once MYBB_ROOT."inc/mailhandlers/smtp.php";
-			$mail = new SmtpMail();
-		}
-		else
-		{
-			require_once MYBB_ROOT."inc/mailhandlers/php.php";
-			$mail = new PhpMail();
-		}
-	}
-
-	// Using SMTP based mail
-	if($mybb->settings['mail_handler'] == 'smtp')
+	// If MyBB's built-in SMTP mail handler is used, set the keep alive bit accordingly.
+	if(isset($mybb->settings['mail_handler']) && $mybb->settings['mail_handler'] == 'smtp' && $keep_alive == true && !empty($mail->keep_alive))
```

Comment:
What if the custom handler set `$mail->keep_alive` to `false`?

Reply:
Hmm, I didn't think about that before. Now I have two perspectives on behalf of your comment:
1. How can a custom mail handler set keep_alive?
2. Since here keep_alive is set to true, what if a custom mail handler is going to set it to false?

I guess the first one is what you mean.

I can't think of a good way to set keep_alive here for custom handlers, as we may not be sure if the custom handler could have a similar member variable `$keep_alive` (so is `build_message` method below before actual sending). Anyway, it can be set via any one of the following two hooks just before sending the mail, though it's not quite intuitive.

Reply:
I'm a little confused, here, `!empty($mail->keep_alive)` seems to check for not false, , which makes it redundant to set to `true` within it (`$mail->keep_alive = true;` below). Perhaps an `else` statement is missing?

Perhaps this block should be moved into `get_my_mailhandler()` ?